common.rkt

```diff
@@ -167,7 +167,7 @@
 (struct Ans (term constraint) #:prefab)
 
 (define (walk* tm st)
-  (let* ((sub (state-sub st)) (tm (walk tm sub)))
+  (let* ((sub (if (state? st) (state-sub st) st)) (tm (walk tm sub)))
```

Comment:
`walk*` doesn't need a state, just a substitution.  Let's change the callers to always pass a substitution.